data_mgr.py

```diff
@@ -0,0 +1,238 @@
+            for n in getattr(k, node_type)]
+        result = []
+        for node in nodes:
+            if self._node_filter(node, args):
+                result.append(node)
+        return result
+
+    def get_nodes_id(self, node_type, args):
+        """Return protobuf node objects for given id."""
+        nat_ids = args.get('native_ids', [])
+        w_ids = []
+        for nat_id in nat_ids:
+            o_name, w_name, nid = nat_id.split('/', 2)
+            w_ids.append(f"{o_name}/{w_name}")
+        if node_type == 'proxy_nodes':
+            nodes = (
+                n for w_id in set(w_ids)
+                for n in (
+                    list(getattr(self.data[w_id], 'task_proxies', []))
+                    + list(getattr(self.data[w_id], 'family_proxies', []))))
+        else:
+            nodes = (
+                n for w_id in set(w_ids)
+                for n in getattr(self.data[w_id], node_type, []))
+        result = []
+        for node in nodes:
+            if ((not nat_ids or node.id in set(nat_ids)) and
+                    self._node_filter(node, args)):
+                result.append(node)
+        return result
+
+    def get_node_id(self, node_type, args):
+        """Return protobuf node object for given id."""
+        n_id = args.get('id')
+        o_name, w_name, remainder = n_id.split('/', 2)
+        w_id = f"{o_name}/{w_name}"
+        if node_type == 'proxy_nodes':
+            nodes = (
+                list(getattr(self.data[w_id], 'task_proxies', []))
+                + list(getattr(self.data[w_id], 'family_proxies', [])))
+        else:
+            nodes = getattr(self.data[w_id], node_type, [])
+        for node in nodes:
+            if node.id == n_id:
+                return node
+        return None
+
+    def get_edges_all(self, args):
+        """Return all workflows protobuf edge objects for given id."""
+        w_args = {
+            'ids': args['workflows'],
+            'exids': args['exworkflows'],
+        }
+        result = []
+        for edge in [
+                n for k in self.get_workflow_msgs(w_args)
+                for n in getattr(k, 'edges')]:
+            result.append(edge)
+        return result
+
+    def get_edges_id(self, args):
+        """Return protobuf edge objects for given id."""
+        nat_ids = args.get('native_ids', [])
+        w_ids = []
+        for nat_id in nat_ids:
+            oname, wname, eid = nat_id.split('/', 2)
+            w_ids.append(f"{oname}/{wname}")
+        w_ids = list(set(w_ids))
+        result = []
+        for edge in [
+                e for k in w_ids
+                for e in getattr(self.data[k], 'edges')]:
+            result.append(edge)
+        return result
+
+    def _node_filter(self, node, args):
+        """Filter nodes based on attribute arguments"""
+        node_id = getattr(node, 'id')
+        slash_count = node_id.count('/')
+        n_id = node_id.split('/', slash_count)
```

Comment:
I think it would be easier to test this part, and also reduce the complexity in `_node_filter` if we moved it to a `_get_node_id` function.

Reply:
done.